src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -222,13 +211,92 @@ public sealed class ReadyToRunCodegenNodeFactory : NodeFactory
                   nameMangler,
                   new ReadyToRunTableManager(context))
         {
-            _importMethods = new Dictionary<TypeAndMethod, IMethodNode>();
 
             Resolver = moduleTokenResolver;
             InputModuleContext = signatureContext;
             CopiedCorHeaderNode = corHeaderNode;
             if (!win32Resources.IsEmpty)
                 Win32ResourcesNode = new Win32ResourcesNode(win32Resources);
+
+            // Create node caches
+            _constructedHelpers = new NodeCache<ReadyToRunHelper, ISymbolNode>(CreateReadyToRunHelperCell);
+
+            _importMethods = new NodeCache<TypeAndMethod, IMethodNode>(CreateMethodEntrypoint);
+
+            _localMethodCache = new NodeCache<TypeAndMethod, MethodWithGCInfo>(key =>
+            {
+                return new MethodWithGCInfo(key.Method.Method, GetSignatureContext());
+            });
+
+            _methodSignatures = new NodeCache<MethodFixupKey, MethodFixupSignature>(key =>
+            {
+                return new MethodFixupSignature(
+                    key.FixupKind,
+                    key.TypeAndMethod.Method,
+                    InputModuleContext,
+                    key.TypeAndMethod.IsUnboxingStub,
+                    key.TypeAndMethod.IsInstantiatingStub
+                );
+            });
+
+            _typeSignatures = new NodeCache<TypeFixupKey, TypeFixupSignature>(key =>
+            {
+                return new TypeFixupSignature(key.FixupKind, key.TypeDesc, InputModuleContext);
```

Comment:
I think this is the gist of the problem. In release / large version bubble mode (i.o.w. with extensive inlining enabled) not all type signatures are pointed to from the input module (they may be pointed to from a method in an external module belonging to the same version bubble that gets inlined into the method being compiled).

Reply:
@trylek In that case, it sounds like the signature context should be part of the key, since we can encode the same signature in different ways depending on which context to use, or am I missing something?
If so, @SrivastavaAnubhav you should be able to fix all instances by including this context signature as part of your key (and make it part of the GetHashCode/Equals)

Reply:
Correct. That's what I basically told Anubhav in our yesterday afternoon chat.